src/consensus.rs

```diff
@@ -840,18 +925,98 @@ impl Consensus {
             // Broadcast the *Commit* Quorum Certificate to all validators
             self.broadcast_net_message(ConsensusNetMessage::Commit(
                 commit_quorum_certificate.clone(),
+                consensus_proposal_hash,
             ))?;
 
             // Process the same locally.
-            let consensus_proposal = std::mem::take(&mut self.bft_round_state.consensus_proposal);
-            self.commit_proposal(consensus_proposal, commit_quorum_certificate)?;
+            self.try_commit_current_proposal(commit_quorum_certificate)?;
         }
         // TODO(?): Update behaviour when having more ?
         Ok(())
     }
 
     /// Message received by follower.
-    fn on_commit(&mut self, commit_quorum_certificate: QuorumCertificate) -> Result<()> {
+    fn on_commit(
+        &mut self,
+        commit_quorum_certificate: QuorumCertificate,
+        proposal_hash_hint: ConsensusProposalHash,
+    ) -> Result<()> {
+        match self.bft_round_state.state_tag {
+            StateTag::Follower => {
+                return self.try_commit_current_proposal(commit_quorum_certificate)
+            }
+            StateTag::Joining => {}
+            _ => bail!("Commit message received while not follower"),
+        };
+
+        // We are joining consensus, try to sync our state.
+        // First find the prepare message to this commit.
+        let Some(proposal_index) = self
+            .bft_round_state
+            .joining
+            .buffered_prepares
+            .iter()
+            .position(|p| p.hash() == proposal_hash_hint)
```

Comment:
OPT:
```suggestion
            .rposition(|p| p.hash() == proposal_hash_hint)
```
Si je me gourre pas, on push les prepare recu. En parcourant la liste from the right, on gagne du temps I guess 
Clairement optionnel :smile:

Reply:
hmm je ne sais pas si le swap_remove qu'on fait après ne vient pas gêner...